Add cuSOLVER path for torch.linalg.lstsq

IvanYashchuk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54725 Added cuBLAS path for torch.linalg.lstsq
* **#57317 Add cuSOLVER path for torch.linalg.lstsq**

This PR implements QR-based least squares solver using geqrf, ormqr, and
triangular_solve operations.

Internal code of triangular_solve was fixed to handle correctly larger
sized rectangular arrays.

Differential Revision: [D28312683](https://our.internmc.facebook.com/intern/diff/D28312683)

IvanYashchuk committed:
Add cuSOLVER path for torch.linalg.lstsq

This PR implements QR-based least squares solver using geqrf, ormqr, and
triangular_solve operations.

Internal code of triangular_solve was fixed to handle correctly larger
sized rectangular arrays.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSOLVER path for torch.linalg.lstsq"


This PR implements QR-based least squares solver using geqrf, ormqr, and
triangular_solve operations.

Internal code of triangular_solve was fixed to handle correctly larger
sized rectangular arrays.

[ghstack-poisoned]

mruberry commented:
@xwang233 and @Lezcano and/or @nikitaved, would you review this, please?

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -667,7 +667,7 @@
 :attr:`driver` chooses the LAPACK/MAGMA function that will be used.
 For CPU inputs the valid values are `'gels'`, `'gelsy'`, `'gelsd`, `'gelss'`.
-For CUDA input, the only valid driver is `'gels'`, which assumes that :attr:`A` is full-rank and `m < n`.
+For CUDA input, the only valid driver is `'gels'`, which assumes that :attr:`A` is full-rank.
```
Thanks!

Lezcano commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -2799,14 +2800,107 @@ static void apply_gels(const Tensor& a, Tensor& b, Tensor& infos) {
+        const_cast<Tensor&>(infos),
+        upper, transpose, conjugate_transpose, unitriangular);
+
+    B.narrow(-2, m, n - m).zero_();
```
This is because `triangular_solve_kernel` writes its output into the first `m` elements of `B`, right? Could you leave a comment explaining this here?

lezcano reviewed (approved):
LGTM! The logic is as clean as it can be. I just left a small comment on a bit that I found slightly more difficult to understand.

Lezcano commented on aten/src/ATen/native/cuda/BatchLinearAlgebraLib.cu:
```diff
@@ -84,9 +84,10 @@ static void apply_triangular_solve(Tensor& A, Tensor& B, bool upper, bool transp
+  auto n = cuda_int_cast(A.size(-1), "n");
   auto nrhs = cuda_int_cast(B.size(-1), "nrhs");
-  auto lda = std::max<int>(1, n);
+  auto lda = std::max<int>(1, m);
```
Are we testing this case? If not, it'd be good to add tests for the rectangular cases.

IvanYashchuk replied:
We don't test it. It's indirectly tested now with the lstsq/gels using it. The public interface through `torch.triangular_solve` doesn't allow rectangular inputs.

xwang233 commented on test/test_linalg.py:
```diff
@@ -212,11 +212,14 @@ def numpy_ref(a, b):
+        # cases m < n are only supported on CPU and for cuSOLVER path on CUDA
         m_l_n_sizes = [(m // 2, m) for m in ms]
-        matrix_sizes = m_ge_n_sizes + (m_l_n_sizes if device == 'cpu' else [])
+        matrix_sizes = m_ge_n_sizes + (m_l_n_sizes if cusolver_available else [])
```
maybe use `(cusolver_available or device == 'cpu')`  to test both?

xwang233 reviewed (approved):
LGTM. Thanks for the PR!

IvanYashchuk committed:
Update on "Add cuSOLVER path for torch.linalg.lstsq"


This PR implements QR-based least squares solver using geqrf, ormqr, and
triangular_solve operations.

Internal code of triangular_solve was fixed to handle correctly larger
sized rectangular arrays.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSOLVER path for torch.linalg.lstsq"


This PR implements QR-based least squares solver using geqrf, ormqr, and
triangular_solve operations.

Internal code of triangular_solve was fixed to handle correctly larger
sized rectangular arrays.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSOLVER path for torch.linalg.lstsq"


This PR implements QR-based least squares solver using geqrf, ormqr, and
triangular_solve operations.

Internal code of triangular_solve was fixed to handle correctly larger
sized rectangular arrays.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Add cuSOLVER path for torch.linalg.lstsq"


This PR implements QR-based least squares solver using geqrf, ormqr, and
triangular_solve operations.

Internal code of triangular_solve was fixed to handle correctly larger
sized rectangular arrays.

[ghstack-poisoned]

mruberry reviewed (approved):
Nice work all! Thanks for reviewing, @nikitaved, @xwang233

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28242069).

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28242069).

samestep commented:
Reverting this PR because it broke one of the Windows test jobs: https://app.circleci.com/pipelines/github/pytorch/pytorch/317376/workflows/463399f8-78ef-4894-a9bf-8b666943efc2/jobs/13217419

mruberry commented:
This diff was revert, but the previous commits in the stack were not, I think. Link to why it was reverted:

https://app.circleci.com/pipelines/github/pytorch/pytorch/317452/workflows/c6beb886-c8ac-4cb5-bdbc-9cf351f89b7c/jobs/13220476

It broke pytorch_windows_vs2019_py36_cuda10.1_test2 and tests test_linalg_lstsq_input_checks_cuda_complex128, test_linalg_lstsq_input_checks_cuda_complex64, test_linalg_lstsq_input_checks_cuda_float32, and test_linalg_lstsq_input_checks_cuda_float64. 

Sample failure snippet:

```
Traceback (most recent call last):
  File "C:\Users\circleci\project\build\win_tmp\build\torch\testing\_internal\common_device_type.py", line 292, in instantiated_test
    result = test_fn(self, *args)
  File "C:\Users\circleci\project\build\win_tmp\build\torch\testing\_internal\common_device_type.py", line 617, in dep_fn
    return fn(slf, device, *args, **kwargs)
  File "C:\Users\circleci\project\build\win_tmp\build\torch\testing\_internal\common_device_type.py", line 617, in dep_fn
    return fn(slf, device, *args, **kwargs)
  File "test_linalg.py", line 401, in test_linalg_lstsq_input_checks
    torch.linalg.lstsq(a, b)
AssertionError: RuntimeError not raised
```

The easiest way to reland the rest of the stack is probably to rebase the uncommitted PRs on nightly with the fix. We can run the updated PR through ci/all to validate this build is fixed, too.

IvanYashchuk committed:
Update on "Add cuSOLVER path for torch.linalg.lstsq"


This PR implements QR-based least squares solver using geqrf, ormqr, and
triangular_solve operations.

Internal code of triangular_solve was fixed to handle correctly larger
sized rectangular arrays.

Differential Revision: [D28242069](https://our.internmc.facebook.com/intern/diff/D28242069)

[ghstack-poisoned]

mruberry commented:
> @mruberry, I fixed the problem with that Windows CUDA 10.1 build. Here is the ci-all PR #57816.
> 
> The problem was that the condition of cuSOLVER availability was not correct in the test. I think we should consider adding a more robust way to check from Python whether cuSOLVER is used in PyTorch.
> 
> We use cuSOLVER if CUDA version is >= **10.1.243**, but `torch._C._cuda_getCompiledVersion()` or `torch.version.cuda` output only major and minor versions (`'10.1'`). The current version of this PR checks that for the case of underdetermined input (`m<n`) the error is thrown for CUDA versions **< 10.1**. However, the underdetermined case is actually not working and raises the error for CUDA versions **< 10.1.243**, so raising the error is not tested for versions from **10.1.0** to **10.1.243**.

Thanks @IvanYashchuk, and thanks for the thorough analysis. So users with a CUDA version between 10.1 and 10.1.243 will get the correct behavior (we think), but our test suite will report the behavior as incorrect?

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28312683).

IvanYashchuk commented:
> So users with a CUDA version between 10.1 and 10.1.243 will get the correct behavior (we think)

Yes, but our test suite doesn't test the behavior for these versions, the tests will pass.
